OSDOCS-13820: Adding Kueue about docs (#91855)
openshift/openshift-docs: welcome/about-kueue.adoc

:_mod-docs-content-type: ASSEMBLY
include::_attributes/common-attributes.adoc[]
[id="about-kueue"]
= About {product-title}
:context: about-kueue

toc::[]

> {product-title} is a Kubernetes-native system that manages quotas and how jobs consume them. Kueue decides when a job waits, when a job is admitted to start by creating pods, and when a job should be _preempted_ (active pods should be deleted).

Reviewer: So Kueue has a concept of an integration (Job, PyTorchJob, JobSet, Pod, Deployment, StatefulSet).

In reality, Kueue can decide when to admit these workloads.

The docs mention Jobs so I know it can be confusing as we expalin the other integrations. It is something to keep in mind..

Kueue takes the approach that "Job" is a general workload and it means any kind of Kueue workload. I'm not sure if that is what we mean here.

Author: probably better to use "workload" in that case, since Job generally means something so specific :)

Kueue is compatible with environments that use heterogeneous, elastic resources. This means that the environment has many different resource types, and those resources are capable of dynamic scaling.

{product-title} is based on the link:https://kueue.sigs.k8s.io/docs/[Kueue] open source project.

// Personas
[id="about-kueue-personas"]
== {kueue-s} personas

Different personas exist in a {kueue-s} workflow.

Batch administrators:: Batch administrators manage the cluster infrastructure and establish quotas and queues.
Batch users:: Batch users run workloads on the cluster. Examples of batch users might be researchers, AI/ML engineers, or data scientists.
Serving users:: Serving users run workloads on the cluster. For example, to expose a trained AI/ML model for inference.
Platform developers:: Platform developers integrate Kueue with other software. They might also contribute to the Kueue open source project.

// How it works
[id="about-kueue-workflow"]
== {kueue-s} workflow
// TODO: add diagram?

> The {kueue-s} workflow can be described at a high level as follows:

Reviewer: This is very specific to the Job controller.

I'm not sure if you want this doc to only mean Kueue is used to manage batch jobs.

And you will expand on this in other sections?

Author: Updated step 5 to make this general

:_mod-docs-content-type: ASSEMBLY
include::_attributes/common-attributes.adoc[]
[id="about-kueue"]
= About {product-title}
:context: about-kueue

toc::[]

> {product-title} is a Kubernetes-native system that manages access to resources for jobs.

Reviewer: ```suggestion
{product-title} is a Kubernetes-native system that manages access to resources for workloads.
```

Author: Discussed offline and decided to agree on "jobs" since the note explaining this is included. We will include more details about different job types as docs are added for other features.

> {product-title} can determine when a job waits, is admitted to start by creating pods, or should be _preempted_, meaning that active pods for that job are deleted.

Reviewer: ```suggestion
{product-title} can determine when a workload waits, is admitted to start by resuming the workload, or should be _preempted_, meaning that active pods for that workload are deleted.
```

[NOTE]
====
In the context of {product-title}, a job can be defined as a one-time or on-demand task that runs to completion.
====

{product-title} is based on the link:https://kueue.sigs.k8s.io/docs/[Kueue] open source project.

{product-title} is compatible with environments that use heterogeneous, elastic resources. This means that the environment has many different resource types, and those resources are capable of dynamic scaling.

{product-title} does not replace any existing components in a Kubernetes cluster, but instead integrates with the existing Kubernetes API server, scheduler, and cluster autoscaler components.

> {product-title} supports all-or-nothing semantics. This means that either an entire change is applied successfully to pods, or the entire change is rolled back, preventing partial updates and issues that could be caused by them.

Reviewer: ```suggestion
{product-title} supports all-or-nothing semantics. This means that a workload can be admitted with all of its components or it would be rejected if the workload can not fit on the cluster.
```

Something like that.

// Personas
[id="about-kueue-personas"]
== Personas

Different personas exist in a {product-title} workflow.

Batch administrators:: Batch administrators manage the cluster infrastructure and establish quotas and queues.
Batch users:: Batch users run jobs on the cluster. Examples of batch users might be researchers, AI/ML engineers, or data scientists.
Serving users:: Serving users run jobs on the cluster. For example, to expose a trained AI/ML model for inference.
Platform developers:: Platform developers integrate {product-title} with other software. They might also contribute to the Kueue open source project.

[id="about-kueue-workflow"]
== Workflow overview
// TODO: add diagram?

The {product-title} workflow can be described at a high level as follows:

> . Batch administrators create and configure `ResourceFlavor`, `LocalQueue`, `ClusterQueue`, `AdmissionCheck`, and `BorrowingCohort` resources.

Reviewer: So right now, I'd maybe mention ResourceFlavor, LocalQueue and ClusterQueue.

AdmissionChecks is optional.

I don't know what BorrowingCohort is but I am not familar with any CRD called that in Kueue.

> . User personas create jobs on the cluster.

Reviewer: ```suggestion
. User personas creates a workload on the cluster.
```

. The Kubernetes API server validates and accepts job data.

> . {product-title} admits jobs based on configured options, such as order or quota. It injects affinity into the job by using resource flavors, and creates a `Workload` object that corresponds to each job.

Reviewer: ```suggestion
. {product-title} admits workloads based on configured options, such as order or quota. It injects affinity into the workload by using resource flavors.
```

IMO workloads are a bit of an implementation detail. Users do not interact with these objects. They can query those objects to see the status of the admission but I'm not sure it belongs in the about-docs.

> . Depending on the job type, one of the available controllers creates pods.

Reviewer: ```suggestion
```
I think all workloads create pods.

Author: lol yeah sorry, what I meant to say here was "the type of job determines which controller creates the pods", but that might not be important. I'll make this clearer.
